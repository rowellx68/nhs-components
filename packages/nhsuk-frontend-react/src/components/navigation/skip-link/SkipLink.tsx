'use client';

import React, { useEffect } from 'react';
import { ElementProps } from '@/types/shared';
import clsx from 'clsx';
import { Factory, factory } from '@/internal/factory/factory';
import initSkipLink from '@/resources/skip-link/skip-link';

export type SkipLinkProps = ElementProps<'a', 'href'>;

type SkipLinkFactory = Factory<{
  props: SkipLinkProps;
  ref: HTMLAnchorElement;
}>;

const SkipLink = factory<SkipLinkFactory>(
  ({ children, className, ...props }, ref) => {
    useEffect(() => {
      setTimeout(() => {
        initSkipLink();
      }, 500);
    });

    return (
      <a
        className={clsx('nhsuk-skip-link', className)}
        href="#maincontent"
        tabIndex={0}
        {...props}
        ref={ref}
      >
        {children}
      </a>
    );
  },
);

SkipLink.displayName = 'SkipLink';

export { SkipLink };
