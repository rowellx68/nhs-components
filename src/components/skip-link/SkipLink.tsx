'use client';

import React, { useEffect, useImperativeHandle, useRef } from 'react';
import { ElementProps } from '@/types/shared';
import clsx from 'clsx';
import { Factory, factory } from '@/internal/factory/factory';
import { SkipLink as NhsSkipLink } from 'nhsuk-frontend';

export type SkipLinkProps = ElementProps<'a'>;

type SkipLinkFactory = Factory<{
  props: SkipLinkProps;
  ref: HTMLAnchorElement;
}>;

const SkipLink = factory<SkipLinkFactory>(
  ({ children, className, href = '#maincontent', ...props }, ref) => {
    const internalRef = useRef<HTMLAnchorElement>(null);
    useImperativeHandle(ref, () => internalRef.current as HTMLAnchorElement);

    useEffect(() => {
      if (!internalRef.current) {
        return;
      }

      new NhsSkipLink(internalRef.current);
    }, [internalRef]);

    return (
      <a
        className={clsx('nhsuk-skip-link', className)}
        href={href}
        tabIndex={0}
        {...props}
        ref={internalRef}
      >
        {children}
      </a>
    );
  },
);

SkipLink.displayName = 'SkipLink';

export { SkipLink };
