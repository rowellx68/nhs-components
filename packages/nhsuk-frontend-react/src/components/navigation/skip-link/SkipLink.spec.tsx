import React from 'react';
import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { SkipLink } from './SkipLink';

it('should render the SkipLink component', () => {
  const { container } = render(<SkipLink>Skip to main content</SkipLink>);

  expect(container).toMatchSnapshot();
});
