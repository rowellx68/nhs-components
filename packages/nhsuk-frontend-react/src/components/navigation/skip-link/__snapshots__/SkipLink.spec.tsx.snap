// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the SkipLink component 1`] = `
<div>
  <a
    class="nhsuk-skip-link"
    href="#maincontent"
  >
    Skip to main content
  </a>
</div>
`;
